### DISCUSSION TITLE
Use `kubectl` to update nodes.longhorn.io (instead of Python client)

### DISCUSSION THREAD

**Discussion by guettli** [longhorn/longhorn#9057] [General]
I see two ways of to add a disk to Longhorn with a script:

* [Python Client](https://longhorn.io/docs/1.6.2/references/longhorn-client-python/)
* [Default disk/node config](https://longhorn.io/docs/1.6.2/nodes-and-volumes/nodes/default-disk-and-node-config/)

I would like to use a third way: `kubectl`

When I add a disk via the UI, then I can clearly see the changes in yaml output of `kubectl get nodes.longhorn.io mynode`.

What could fail, if we update the node-object via `kubectl` or a similar tool?

**Comment by ejweber** [comment]
I would add an existing third way to your list, which is quite commonly used:

- [Longhorn UI](https://longhorn.io/docs/1.6.2/nodes-and-volumes/nodes/multidisk/#add-a-disk)

As for your question, I can't think of any particular reason not to use `kubectl` as you describe. When you use the UI or Python client, you are hitting the `DiskUpdate` API. The code for that API is quite simple:

- https://github.com/longhorn/longhorn-manager/blob/6fc75afd54aa532f34b8322e268afc91cc1ee952/api/node.go#L95-L121
- https://github.com/longhorn/longhorn-manager/blob/6fc75afd54aa532f34b8322e268afc91cc1ee952/manager/node.go#L104-L118

We simply read the new map of DiskSpecs and update the Longhorn Node CR accordingly. The webhook does some validation on the new map regardless of what method you choose.

https://github.com/longhorn/longhorn-manager/blob/6fc75afd54aa532f34b8322e268afc91cc1ee952/webhook/resources/node/validator.go#L63-L79

Some warnings:

- It seems we do at least some validation inside the Longhorn UI when adding disks. (For example, we do not allow you to add a new disk with the same name as an old disk. I tried this with the `kubectl edit` approach and the old disk spec was overwritten. This would almost certainly lead to some chaos.) The `kubectl` approach circumvents this validation.
- Edits should be made to the `nodes.longhorn.io` objects and not the Kubernetes `nodes` objects. (This is probably obvious.)
- Edits should never be made to node/disk status.

**Reply by guettli** [reply]
@ejweber thank you very much for your reply. We already use `kubectl` in our automation (the UI is no option for is in this case). Good to know that at least some validation gets done in a webhook.
